lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs

```diff
@@ -98,28 +98,48 @@ public void SetException(Exception e)
         {
             lock (_heartBeatManager)
             {
-                try
+                if (_lastException.IsPresent())
                 {
-                    if (!_lastException.IsPresent())
-                    {
-                        _lastException = Optional<Exception>.Of(e);
-                    }
+                    return;
+                }
 
-                    State = TaskState.Failed;
+                _lastException = Optional<Exception>.Of(e);
+                State = TaskState.Failed;
+
+                try
+                {
                     _taskLifeCycle.Stop();
                 }
-                finally
+                catch (Exception stopException)
                 {
-                    Heartbeat();
+                    Utilities.Diagnostics.Exceptions.Caught(
```

Comment:
Right. The Java side does this by making the default implementation in `org.apache.reef.runtime.common.evaluator.task.defaults.DefaultCloseHandler` throw a `RuntimeException`. This kills the Evaluator. That might be a bit drastic, but there is really no way for us to kill just the task, as it might have spawned threads.